drift/core/psmc.py

```diff
@@ -201,6 +206,356 @@ def _work_fisher_bias_m(self, mi):
+        """A method to estimate the noise bias. If noise=True then the q-estimator MPI-array is one longer, and the last parameter is the projection against the noise.
+
+        Parameters
+        ----------
+        mi : integer
+            The m-mode we are calculating for.
+        vec1, vec2 : np.ndarrays[num_kl, num_realisatons]
+            The vector(s) of data we are estimating from. These are KL-mode coefficients. Passing in `vec2` is optional and for cross power spectrum calculation.
+        """
+        # Calculate q_a for noise power (x0^H N x0 = |x0|^2)
+        evals, evecs = self.kltrans.modes_m(mi)
+
+        if evals is not None:
+
+            # If calculating crosspower don't include instrumental noise
+            noisemodes = 0.0 if self.crosspower else 1.0
+            noisemodes = noisemodes + (evals if self.zero_mean else 0.0)
+
+            x0 = (vec1.T / (evals + 1.0)).T
+
+            if vec2 is not None:
+                y0 = (vec2.T / (evals + 1.0)).T
+            else:
+                y0 = x0
+
+            self.qa.global_slice[mi, -1] = np.sum(
+                (x0 * y0.conj()).T.real * noisemodes, axis=-1
+            )
+
+    def recv_send_data(self, data, axis):
+        """Send data from one MPI process to the next by using a vector buffer.
+
+        Parameters
+        ----------
+        data : np.ndarray, complex
+            Data vector that is sent to the next rank. Expecting
+            a complex data type.
+        axis : int
+            Axis over which the data is devided into chunks of 4GB.
+
+        Returns
+        -------
+        recv_buffer : np.ndarray
+            The data vector received from the previous rank.
+        """
+
+        comm = MPI.COMM_WORLD
+        size = comm.Get_size()
+        rank = comm.Get_rank()
+
+        shape = data.shape
+        dtype = data.dtype
+
+        recv_rank = (rank - 1) % size
+        send_rank = (rank + 1) % size
+
+        recv_buffer = np.zeros(shape, dtype=dtype)
+
+        # Need to send in 4GB chunks due to some MPI library.
+        message_size = 4 * 2 ** 30.0
+        dsize = np.prod(shape) * 16.0
+        num_messages = int(np.ceil(dsize / message_size))
+
+        num, sm, em = mpiutil.split_m(shape[axis], num_messages)
+
+        for i in range(num_messages):
+            slc = [slice(None)] * len(shape)
+            slc[axis] = slice(sm[i], em[i])
+
+            di = np.ascontiguousarray(data[slc], dtype=dtype)
+            bi = np.zeros(di.shape, dtype=dtype)
+
+            # Initiate non-blocking receive
+            request = comm.Irecv(
+                [bi, MPI.DOUBLE_COMPLEX], source=recv_rank, tag=(i * size) + recv_rank
+            )
+            # Initiate send
+            comm.Send([di, MPI.DOUBLE_COMPLEX], dest=send_rank, tag=(i * size) + rank)
+            # Wait for receive
+            request.Wait()
+
+            # Fill recv buffer with messages
+            recv_buffer[slc] = bi
```

Comment:
oh yeah, I have been getting a warning that this should be recv_buffer[tuple(slc)] instead of recv_buffer[slc]. Any opinions?